### ISSUE TITLE
More literals for Morganey

### ISSUE THREAD

**Issue by keddelzz** [morganey-lang/Morganey#58]
As you know, I started working on implementing a interpolator for lambda terms (#25). 

In the scope of the issue I played with interoperability between Morganey and Scala. For that it would be nice to make Morganey support other literals as well: char-literals, string-literals, list-literals. This would make the extension for the lambda calculus, which is already included in Morganey (i.e.: int-literals and name bindings), even more useful.

But I see a problem: 
String-literals and list-literals can easily generate big programs, because they expand to their eqivalent in church-encoding. Those big programs are unsafe if used in horta-hell.
For the usage in horta literals might be too unsafe anyway: int literals can also generate huge data-structures.
## 

So I don't know if you agree with me (to implement more literals) or if you consider them too unsafe. Nevertheless it would be possible to only allow them in the use of the interpolator. This would have the advantage, that they'll be not usable in horta, but they will generate a nice interoperability between Scala and Morganey.

**Comment by rexim** [comment]
The truth is I've never wanted Morganey to be a real useful programming language. The intension was to create yet another Turing tarpit, embed it to a chat bot and see people play with it. But after adding number literals #37 and bindings #23 it already looks too "useful" for me to actually serve the initial purposes. :D So, hey! Why not? Let's do this! Let's see how far we can go! Maybe we can even implement that on this week Sunday. ;)

As for the bloating IR, I have a couple of ideas on how to solve that problem. One of them is to introduce a different kind of internal Application that stores left term, right term, and a number which represents how much times the left term was applied to the right one. This may probably help with representing the numbers. We can replace `(\f . (\x . (f (f (f (f (f x)))))))` with `(\f . (\x . [f x 5]))`. And during reduction we should treat such "collapsed" Application as series of Applications. That may really help with big numbers. I'll create an issue for all of that.

As for the rest of the structures, I'll think about that and create issues as soon as I have something.

Thanks!

**Comment by keddelzz** [comment]
I've thought about the different representation, too. 
This might be also possible with what i've planned for #25. Maybe we could have a short chat about that later. But as a short introduction for my idea:

``` scala
// extract values from lamba terms to scala
val m"( ${a: Int} ${b: Int} )" = someLambdaTerm()

// calulate it in scala
val res = math.pow(a, b)

// interpolate back to terms
val term = m"$res"
```

This is totally doable and already planned for #25.

Edit: But this does only effect the evaluation. It will still generate the bloated IR.

**Comment by rexim** [comment]
Wow! Looks awesome! So it's sorta lambda term destructuring. :) I think I see what you mean here. Right now we already know how to detect numbers and pairs in terms. But we can also detect arithmetic operations over those numbers! That will enable us to perform them on the Scala side (which is incredibly faster) rather than with pure reduction. Do I understand you correctly? I like this idea! :)

As for the chat, I'd prefer to keep all of the Morganey related conversations in the issue tracker just to not lose anything important.

**Comment by keddelzz** [comment]
Yes, you do understand me correctly. Though the detection of complex constructs should be much more easy with the "lambda term destruction" as it is purely with lambda terms.

I understand, that you want to keep all Morganey related conversations in the issue tracker. :)

**Comment by rexim** [comment]
- Implemented string literals in #64
- Created separate issue for char literals #65 
- Created separate issue for list literals #66 

Closed this issue because its scope is exhausted. Everything is either implemented or has a separate issue.
